tensorly/decomposition/_nn_cp.py

```diff
@@ -291,7 +294,145 @@ def non_negative_parafac(tensor, rank, n_iter_max=100, init='svd', svd='numpy_sv
+        Default: False
+    verbose: boolean
+        Indicates whether the algorithm prints the successive
+        reconstruction errors or not
+        Default: False
+    return_errors: boolean
+        Indicates whether the algorithm should return all reconstruction errors
+        and computation time of each iteration or not
+        Default: False
+    cvg_criterion : {'abs_rec_error', 'rec_error'}, optional
+       Stopping criterion for ALS, works if `tol` is not None. 
+       If 'rec_error',  ALS stops at current iteration if ``(previous rec_error - current rec_error) < tol``.
+       If 'abs_rec_error', ALS terminates when `|previous rec_error - current rec_error| < tol`.
+    sparsity : float or int
+
+    Returns
+    -------
+    factors : ndarray list
+            list of positive factors of the CP decomposition
+            element `i` is of shape ``(tensor.shape[i], rank)``
+    errors: list
+        A list of reconstruction errors at each iteration of the algorithm.
+
+    References
+    ----------
+    [1]: N. Gillis and F. Glineur, Accelerated Multiplicative Updates and
+    Hierarchical ALS Algorithms for Nonnegative Matrix Factorization,
+    Neural Computation 24 (4): 1085-1105, 2012.
+    """
+
+    weights, factors = initialize_nn_cp(tensor, rank, init=init, svd=svd,
+                                     random_state=None,
+                                     normalize_factors=False)
+
+    norm_tensor = tl.norm(tensor, 2)
+
+    n_modes = tl.ndim(tensor)
+    if sparsity_coefficients is None or isinstance(sparsity_coefficients, float):
+        sparsity_coefficients = [sparsity_coefficients] * n_modes
+
+    if fixed_modes is None:
+        fixed_modes = []
+
+    # Avoiding errors
+    for fixed_value in fixed_modes:
+        sparsity_coefficients[fixed_value] = None
+
+    # Generating the mode update sequence
+    modes = [mode for mode in range(n_modes) if mode not in fixed_modes]
+
+    # initialisation - declare local varaibles
+    rec_errors = []
+
+    # Iteratation
+    for iteration in range(n_iter_max):
+        # One pass of least squares on each updated mode
+        for mode in modes:
+
+            # Computing Hadamard of cross-products
+            pseudo_inverse = tl.tensor(tl.ones((rank, rank)), **tl.context(tensor))
+            for i, factor in enumerate(factors):
+                if i != mode:
+                    pseudo_inverse = pseudo_inverse*tl.dot(tl.transpose(factor), factor)
+
+            if not iteration and weights is not None:
+                # Take into account init weights
+                mttkrp = unfolding_dot_khatri_rao(tensor, (weights, factors), mode)
+            else:
+                mttkrp = unfolding_dot_khatri_rao(tensor, (None, factors), mode)
+
+            # Call the hals resolution with nnls, optimizing the current mode
+            factors[mode] = tl.transpose(
```

Comment:
I think it would be clearer to have this in two lines, to avoid the potentially confusing indexing of the tuple returned by hals_nnls:
```python
nn_factor, _, _ = hals_nnls(...)
factors[mode] = tl.transpose(nn_factor)
```